dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,758 @@
+{
+    if (offset.ring_sizes[offset.current_ring] == 0)
+        addNewRing(data, offset);
+
+    auto & last_polygon = data.dest.back();
+    auto & last_ring = (last_polygon.inners().empty() ? last_polygon.outer() : last_polygon.inners().back());
+    last_ring.push_back(pt);
+
+    --offset.ring_sizes[offset.current_ring];
+}
+
+void addNewPointFromSimplePolygon(const IPolygonDictionary::Point & pt, Data & data, Offset & offset)
+{
+    if (offset.polygon_sizes[offset.current_polygon] == 0)
+    {
+        ++offset.current_polygon;
+        if (offset.current_polygon == offset.polygon_sizes.size())
+            throw Exception{"Incorrect polygon formatting", ErrorCodes::BAD_ARGUMENTS};
+
+    }
+    auto & last_polygon = data.dest.back();
+    last_polygon.outer().push_back(pt);
+    --offset.polygon_sizes[offset.current_polygon];
+}
+
+const IColumn * unrollMultiPolygons(const ColumnPtr & column, Offset & offset)
+{
+    const auto ptr_multi_polygons = typeid_cast<const ColumnArray*>(column.get());
+    if (!ptr_multi_polygons)
+        throw Exception{"Expected a column containing arrays of polygons", ErrorCodes::TYPE_MISMATCH};
+    offset.multi_polygon_sizes.assign(ptr_multi_polygons->getOffsets());
+    makeDifferences(offset.multi_polygon_sizes);
+
+    const auto ptr_polygons = typeid_cast<const ColumnArray*>(&ptr_multi_polygons->getData());
+    if (!ptr_polygons)
+        throw Exception{"Expected a column containing arrays of rings when reading polygons", ErrorCodes::TYPE_MISMATCH};
+    offset.polygon_sizes.assign(ptr_polygons->getOffsets());
+    makeDifferences(offset.polygon_sizes);
+
+    const auto ptr_rings = typeid_cast<const ColumnArray*>(&ptr_polygons->getData());
+    if (!ptr_rings)
+        throw Exception{"Expected a column containing arrays of points when reading rings", ErrorCodes::TYPE_MISMATCH};
+    offset.ring_sizes.assign(ptr_rings->getOffsets());
+    makeDifferences(offset.ring_sizes);
+
+    return ptr_rings->getDataPtr().get();
+}
+
+const IColumn * unrollSimplePolygons(const ColumnPtr & column, Offset & offset)
+{
+    const auto ptr_polygons = typeid_cast<const ColumnArray*>(column.get());
+    if (!ptr_polygons)
+        throw Exception{"Expected a column containing arrays of points", ErrorCodes::TYPE_MISMATCH};
+    offset.polygon_sizes.assign(ptr_polygons->getOffsets());
+    return ptr_polygons->getDataPtr().get();
+}
+
+void getPointsReprByArrays(const IColumn * column, std::vector<IPolygonDictionary::Point> & dest)
+{
+    const auto ptr_points = typeid_cast<const ColumnArray*>(column);
+    const auto ptr_coord = typeid_cast<const ColumnVector<Float64>*>(&ptr_points->getData());
+    if (!ptr_coord)
+        throw Exception{"Expected coordinates to be of type Float64", ErrorCodes::TYPE_MISMATCH};
+    const auto & offsets = ptr_points->getOffsets();
+    for (size_t i = 0; i < offsets.size(); ++i)
+    {
+        if (offsets[i] - (i == 0 ? 0 : offsets[i - 1]) != 2)
```

Comment:
Difficult to read.
I prefer something like:
```c++
UInt64 prev_offset = 0;
for (size_t i = 0; i < offsets.size(); ++i)
{
    if (offsets[i] - prev_offset != 2)
        ...
    ...
    prev_offset = offsets[i];
}
```
But also you can use `offsets[-1]`. It's 0 (we have special implementation in PODArray for this)

Reply:
The first option is probably more readable, done.